src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1188,491 @@ where
+            .map_err(|_| QueryError::Error {
+                message: "failed to batch insert children hashes".to_owned(),
+            })?
+            .iter()
+            .map(|r| r.get(0))
+            .collect();
+
+        if nodes_hash_ids.len() != nodes.len() {
+            // This should not happen if the upsert was successful
+            return Err(QueryError::Error {
+                message: "number of hash ids do not match the number of nodes".to_owned(),
+            });
+        }
+
+        for (h, n) in nodes_hash_ids.into_iter().zip(nodes.iter_mut()) {
+            n.created = block_number;
+            n.hash_id = h;
+        }
+
+        let (params, batch_stmt) = Node::build_batch_insert(name, &nodes);
+
+        // Batch insert all the child hashes
+        let rows_inserted =
+            txn.client
+                .query(&batch_stmt, &params)
+                .await
+                .map_err(|_| QueryError::Error {
+                    message: "failed to batch insert children hashes".to_owned(),
+                })?;
+
+        if rows_inserted.len() != path.len() {
+            return Err(QueryError::Error {
+                message: "failed to insert all merkle nodes".to_string(),
+            });
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<Types, State> MerklizedStateDataSource<Types, State> for SqlStorage
+where
+    Types: NodeType,
+    State: MerklizedState<Types> + 'static,
+    State::Index: ToTraversalPath<State::Arity> + DeserializeOwned + Send + Sync + FromStr,
+    State::Element: DeserializeOwned + Send + Sync,
+    State::NodeValue: Send,
+    State::Commitment: Send,
+{
+    /// Retreives a Merkle path from the database
+    async fn get_path(
+        &self,
+        state_type: &'static str,
+        tree_height: usize,
+        header_state_commitment_field: &'static str,
+        snapshot: Snapshot<State>,
+        key: String,
+    ) -> QueryResult<MerklePath<State::Element, State::Index, State::NodeValue>> {
+        let key = key.parse::<State::Index>().map_err(|_| QueryError::Error {
+            message: "failed to parse key to Index".to_owned(),
+        })?;
+
+        // Get the traversal path of the index
+        let path = State::Index::to_traversal_path(&key, tree_height)
+            .into_iter()
+            .map(|x| x as i64)
+            .collect::<Vec<_>>();
+
+        let (created, merkle_commitment) = match snapshot {
+            Snapshot::<State>::Commit(commit) => {
+                let mut writer = Vec::new();
+                commit
+                    .serialize_compressed(&mut writer)
+                    .map_err(ParseError::Deserialize)?;
+
+                // Get the block height using the merkle commitment.
+                let query = self
+                    .query_one(
+                        &format!(
+                            "SELECT created FROM Header where 
+                             h.data->>{header_state_commitment_field}"
```

Comment:
Should be `height` instead of `created` I think, and looks like we're missing an `= $1` or something?